### DISCUSSION TITLE
[Projects Beta] Calculated Fields / Sums

### DISCUSSION THREAD

**Discussion by tsteckenborn** [community/community#6080] [Projects and Issues]
Calculated fields and sums, in my opinion, could be an interesting addition.

Here are some ideas:

1. Validity and input: Some values can be derived from others. The effort estimate of a developer could be summarized (e.g., into T-shirt sizes) and/or calculated from several estimates (Optimal, Nominal, Pessimistic). Calculating would reduce the amount of input and error sources.

2. Management: Particularly in the table view, sums could be very useful. If I have a numeric field (such as for effort estimates), then the question of how to sum the estimated effort (and potentially remaining resources) quickly arises.

"Input Fields" and "Calculated Fields" could be useful for automating information like percentages and the like without requiring human input.

**Comment by mfsbo** [comment]
Thanks for putting this out. I always miss the project/organisation statistics page and use different tools to get these statistics out. If we have Statistics about a view that would solve this problem. Its very subjective problem but if github projects can setup some basic statistics rule it will help a lot of people. This can be at organisation level (cross repository), repository, project or even project view level. 

These are all just group of issues so any field added to a group of issues can produce a summary statistics page. So below is an example of use case from vscode repo

### Statistics about group of issues
I like how VS Code repo does their game plan with issues mentioned in the game plan. This already summarise the number of issues mentioned on top of the game plan for example https://github.com/microsoft/vscode/issues/132467 mentions 
![image](https://user-images.githubusercontent.com/50488401/136520322-f176e69e-9e53-42f7-9289-b78e30c750de.png)

If any of these issues have a number field that is assigned a value then it would be good to know sum/average/min/max of that number in some kind of summary. As an example
99 tasks have 3 custom fields (just to cover all those different scenarios) 
 - Points: Story/Task Estimation points, 
 - Rating: Rating from users to assess likes or need etc, 
 - Priority: Might be internal priority in Development

Now on Projects user enters these fields or project template contain them. Then some kind of automation maybe updating these fields. and we get to a stage where we have
- 20 issues that have Points set to some number and rest of 79 have no points setup
- 99 of them have rating setup between a value of 1-5
- 30 have priority setup to 1 and 20 have priority setup to 2

It would be good to be able to setup a summary area where 
- we can choose a field like Points and say
  - show total points
  - ignore empty ones in average or include them as 0/default value.
  - show histogram or group them by value

for example it would be good to know
 - Total/Average Points for a given set of issues (view filtered or not). 
 - Min of the points and Max of the points in one project view. 
 - Average of Rating
 - Min and Max of Rating
 - Count of Priorities (this I think can be achieved by grouping and can be brought into statistics for projects page)

### Example of some expectations
On a public repo like vscode it might be good to allow user to share one of these statistical widget to public similar to badge for example on vscode
- A histogram of this month planned items would be useful for public to know if they are getting many high rated items
- A count of high priority issues would help summarise the notes on release or wiki to say how many high priority items have been closed during certain time period.

Hope this open a lot of opportunities for all of us.

**Comment by 2percentsilk** [comment]
We've just shipped [sum of numeric fields](https://github.blog/changelog/2022-09-27-the-new-github-issues-september-27th-update/). This is just the first step to address many of the items you've listed here but hopefully a step in the right direction to address your needs.

**Comment by itsBryantP** [comment]
I'd like to also have the ability to have custom fields that are formulas based on the value of other numeric fields. An example would be: I'd like to have a field called WSJF (weighted shorted job first) which has a fomula based on 
```
WSJF = (business value + time criticality + Risk) / Job Duration
```
all of which are relative numeric values in other custom fields.

**Reply by zaneulhaq** [reply]
Agreed, WSJF would be extremely helpful. Surely this isn't hard to implement - same as RICE concept. Please advise?

**Comment by cedricziel** [comment]
I would like calculated fields to introduce some simple scoring f.e. with RICE:

```
(Reach x Impact x Confidence) / Effort = Score
```

**Reply by mtstaffa** [reply]
I would like this too.

**Comment by widal001** [comment]
While most of the suggestions have focused on numeric calculations, I might also suggest calculated fields with other data types as well. One that would be helpful for our use cases is being able to dynamically calculate "on time" vs "late" status:

```
deadline:<@today -status:done
```

Currently we have a few workarounds to achieve this goal:

- A custom view that uses filters to identify late tasks
- A custom GitHub Action that updates a dropdown field to reflect this status

Ideally this field could be calculated dynamically and viewed within existing views rather than a separate view.

**Comment by rainbow1016** [comment]
very good question!

**Comment by bbrzoska** [comment]
We'd like to be able to write a custom JavaScript/TypeScript function for the formula, with inputs being all other non-computed fields. We'd like to be able to order/calculate start/end dates for items based on other fields, and our algorithm is complex. Using a proper programming language to write custom formulas would be ideal.

**Reply by mfsbo** [reply]
Can you give some use cases on why we need javascript/typescript.

**Reply by bbrzoska** [reply]
Yes, right now we have a spreadsheet with [custom functions to calculate the roadmap](https://github.com/niieani/google-sheets-roadmap-plan/blob/8da81c7a3d3e720148c23956a2d677d472ac392a/src/roadmap.ts#L74-L81) (start and end dates) for each item. It's pretty complex. Simple formulas aren't enough.

**Comment by RemyDuijkeren** [comment]
I also have a need to calculate the Prio (number) from the fields Effort (options) and Value (options).

I think a better solution would be to handle (complex) calculation by GitHub Actions. Allow us to trigger based on fields and to update fields in GitHub Actions.

In GitHub Actions you can make it as easy or complex as you need.

**Comment by hungluu** [comment]
Having computed fields is the only way we can escape excel, google sheet and other project management tools to all-in into Github project.

**Comment by teohsinyee** [comment]
Any update on this? Need this to calc engineering team DORA within projects

**Comment by Lindbergo** [comment]
+1

**Comment by hungluu** [comment]
Upppppp, still waiting for years until I can completely migrate every planning files / tools to github project 👍🏻. Some of my usual use cases:
- Sum of a range of cells or just some cell, example formula: `SUM(B1:G1)`
- Reference of another cell, for e.g actual completion date of phase 1 is C3, so start date of phase 2 is D3 `=C3 + 1d`
- Selecting some `number` cells show their SUM on header or footer bar
